src/group_theory/perm/concrete_cycle.lean

```diff
@@ -0,0 +1,543 @@
+  form_perm s h.nodup = 1 :=
+begin
+  induction s using quot.induction_on,
+  simp only [form_perm_coe, mk_eq_coe],
+  simp only [length_subsingleton_iff, length_coe, mk_eq_coe] at h,
+  cases s with hd tl,
+  { simp },
+  { simp only [length_eq_zero, add_le_iff_nonpos_left, list.length, nonpos_iff_eq_zero] at h,
+    simp [h] }
+end
+
+lemma is_cycle_form_perm (s : cycle α) (h : nodup s) (hn : nontrivial s) :
+  is_cycle (form_perm s h) :=
+begin
+  induction s using quot.induction_on,
+  exact list.is_cycle_form_perm h (length_nontrivial hn)
+end
+
+lemma support_form_perm [fintype α] (s : cycle α) (h : nodup s) (hn : nontrivial s) :
+  support (form_perm s h) = s.to_finset :=
+begin
+  induction s using quot.induction_on,
+  refine support_form_perm_of_nodup s h _,
+  rintro _ rfl,
+  simpa [nat.succ_le_succ_iff] using length_nontrivial hn
+end
+
+lemma form_perm_apply_not_mem (s : cycle α) (h : nodup s) (x : α) (hx : x ∉ s) :
+  form_perm s h x = x :=
+begin
+  induction s using quot.induction_on,
+  simpa using list.form_perm_apply_not_mem _ _ hx
+end
+
+lemma form_perm_apply_mem (s : cycle α) (h : nodup s) (x : α) (hx : x ∈ s) :
+  form_perm s h x = next s h x hx :=
+begin
+  induction s using quot.induction_on,
+  simpa using list.form_perm_apply_mem_eq_next h _ _
+end
+
+lemma form_perm_reverse (s : cycle α) (h : nodup s) :
+  form_perm s.reverse (nodup_reverse_iff.mpr h) = (form_perm s h)⁻¹ :=
+begin
+  induction s using quot.induction_on,
+  simpa using form_perm_reverse _ h
+end
+
+lemma form_perm_eq_form_perm_iff {α : Type*} [decidable_eq α]
+  {s s' : cycle α} {hs : s.nodup} {hs' : s'.nodup} :
+  s.form_perm hs = s'.form_perm hs' ↔ s = s' ∨ s.subsingleton ∧ s'.subsingleton :=
+begin
+  rw [cycle.length_subsingleton_iff, cycle.length_subsingleton_iff],
+  revert s s',
+  intros s s',
+  apply quotient.induction_on₂' s s',
+  intros l l',
+  simpa using form_perm_eq_form_perm_iff
+end
+
+end cycle
+
+variables {α : Type*}
+
+namespace equiv.perm
+
+variables [fintype α] [decidable_eq α] (p : equiv.perm α) (x : α)
+
+/--
+Given `f : perm α` and `x : α`, generate the list `[x, f x, f (f x), ...]`
+until looping. That means when `f x = x`, `to_list f x = []`.
+-/
+def to_list : list α :=
+(list.range (cycle_of p x).support.card).map (λ k, (p ^ k) x)
+
+@[simp] lemma to_list_one : to_list (1 : perm α) x = [] :=
+by simp [to_list, cycle_of_one]
+
+@[simp] lemma to_list_eq_nil_iff {p : perm α} {x} : to_list p x = [] ↔ x ∉ p.support :=
+by simp [to_list]
+
+@[simp] lemma length_to_list : length (to_list p x) = (cycle_of p x).support.card :=
+by simp [to_list]
+
+lemma to_list_ne_singleton (y : α) : to_list p x ≠ [y] :=
+begin
+  intro H,
+  simpa [card_support_ne_one] using congr_arg length H
+end
+
+lemma two_le_length_to_list_of_mem_support (h : x ∈ p.support) : 2 ≤ length (to_list p x) :=
+by simpa using is_cycle.two_le_card_support (is_cycle_cycle_of _ (mem_support.mp h))
```

Comment:
This is an `iff`, right? It might make a good `simp` lemma for `simpa`ing away some of the `form_perm` hypotheses.
```suggestion
@[simp]
lemma two_le_length_to_list_of_mem_support_iff : 2 ≤ length (to_list p x) ↔ x ∈ p.support :=
sorry
```